eventuate-local-java-cdc-connector-common/src/main/java/io/eventuate/local/common/CdcDataPublisher.java

```diff
@@ -76,42 +85,38 @@ public void handleEvent(EVENT publishedEvent) throws EventuateLocalPublishingExc
 
     String aggregateTopic = publishingStrategy.topicFor(publishedEvent);
 
-    Exception lastException = null;
-
-    for (int i = 0; i < 5; i++) {
-      try {
-        if (publishedEvent.getBinlogFileOffset().map(o -> publishingFilter.shouldBePublished(o, aggregateTopic)).orElse(true)) {
-          logger.info("sending record: {}", json);
-          producer.send(
-                  aggregateTopic,
-                  publishingStrategy.partitionKeyFor(publishedEvent),
-                  json
-          ).get(10, TimeUnit.SECONDS);
-
-          logger.info("record sent: {}", json);
-          lastMessagePublishingFailed = false;
-
-          publishingStrategy.getCreateTime(publishedEvent).ifPresent(time -> distributionSummaryEventAge.record(System.currentTimeMillis() - time));
-          meterEventsPublished.increment();
-        } else {
-          logger.debug("Duplicate event {}", publishedEvent);
-          meterEventsDuplicates.increment();
-        }
-        return;
-      } catch (Exception e) {
-
-        lastMessagePublishingFailed = true;
-        logger.warn("error publishing to " + aggregateTopic, e);
-        meterEventsRetries.increment();
-        lastException = e;
-
-        try {
-          Thread.sleep((int) Math.pow(2, i) * 1000);
-        } catch (InterruptedException ie) {
-          throw new RuntimeException(ie);
-        }
-      }
+    CompletableFuture<Object> result = new CompletableFuture<>();
+
+    if (publishedEvent.getBinlogFileOffset().map(o -> publishingFilter.shouldBePublished(o, aggregateTopic)).orElse(true)) {
+      logger.info("sending record: {}", json);
+
+      send(publishedEvent, aggregateTopic, json, new AtomicInteger(0), result);
+
+      return result;
+    } else {
+      logger.debug("Duplicate event {}", publishedEvent);
+      result.complete(null);
+      return result;
     }
-    throw new EventuateLocalPublishingException("error publishing to " + aggregateTopic, lastException);
+  }
+
+  private void send(EVENT publishedEvent, String aggregateTopic, String json, AtomicInteger retries, CompletableFuture<Object> result) {
```

Comment:
In addition to `totallyProcessedEvents`, it would be useful to `send()` to measure how long the send takes.

Reply:
ok, I will add this measurement